python/ray/tune/experimental/output.py

```diff
@@ -0,0 +1,585 @@
+# Note this is ordered!
+DEFAULT_COLUMNS = collections.OrderedDict(
+    {
+        MEAN_ACCURACY: "acc",
+        MEAN_LOSS: "loss",
+        TRAINING_ITERATION: "iter",
+        TIME_TOTAL_S: "total time (s)",
+        TIMESTEPS_TOTAL: "ts",
+        EPISODE_REWARD_MEAN: "reward",
+    }
+)
+
+VALID_SUMMARY_TYPES = {
+    int,
+    float,
+    np.float32,
+    np.float64,
+    np.int32,
+    np.int64,
+    type(None),
+}
+
+
+class AirVerbosity(Enum):
+    SILENT = 0
+    DEFAULT = 1
+    VERBOSE = 2
+
+    def __lt__(self, b):
+        return self.value < b.value
+
+
+try:
+    class_name = get_ipython().__class__.__name__
+    IS_NOTEBOOK = True if "Terminal" not in class_name else False
+except NameError:
+    IS_NOTEBOOK = False
+
+
+def get_air_verbosity() -> Optional[AirVerbosity]:
+    verbosity = os.environ.get("AIR_VERBOSITY", None)
+    if verbosity:
+        return AirVerbosity(int(verbosity)) if verbosity else None
+
+
+def _get_current_time():
+    return datetime.datetime.fromtimestamp(time.time())
+
+
+def _get_time_str(start_time: float, current_time: float) -> Tuple[str, str]:
+    """Get strings representing the current and elapsed time.
+
+    Args:
+        start_time: POSIX timestamp of the start of the tune run
+        current_time: POSIX timestamp giving the current time
+
+    Returns:
+        Current time and elapsed time for the current run
+    """
+    current_time_dt = datetime.datetime.fromtimestamp(current_time)
+    start_time_dt = datetime.datetime.fromtimestamp(start_time)
+    delta: datetime.timedelta = current_time_dt - start_time_dt
+
+    rest = delta.total_seconds()
+    days = rest // (60 * 60 * 24)
+
+    rest -= days * (60 * 60 * 24)
+    hours = rest // (60 * 60)
+
+    rest -= hours * (60 * 60)
+    minutes = rest // 60
+
+    seconds = rest - minutes * 60
+
+    if days > 0:
+        running_for_str = f"{days:.0f} days, "
+    else:
+        running_for_str = ""
+
+    running_for_str += f"{hours:02.0f}:{minutes:02.0f}:{seconds:05.2f}"
+
+    return f"{current_time_dt:%Y-%m-%d %H:%M:%S}", running_for_str
+
+
+def _get_trials_by_state(trials: List[Trial]) -> Dict[str, List[Trial]]:
+    trials_by_state = collections.defaultdict(list)
+    for t in trials:
+        trials_by_state[t.status].append(t)
+    return trials_by_state
+
+
+def _infer_user_metrics(trials: List[Trial], limit: int = 4) -> List[str]:
+    """Try to infer the metrics to print out."""
+    result = list()
+    for t in trials:
+        if not t.last_result:
+            continue
+        for metric, value in t.last_result.items():
+            if metric not in DEFAULT_COLUMNS:
+                if metric not in AUTO_RESULT_KEYS:
+                    if type(value) in VALID_SUMMARY_TYPES:
+                        result.append(metric)
+
+            if len(result) >= limit:
+                return result
+    return result
+
+
+def _current_best_trial(trials: List[Trial], metric: str, mode: str):
```

Comment:
actually, why do we duplicate the logics here?
I imagine Tune must already be computing this somewhere?

Reply:
yes, some of the logic is duplicated with current `progress_reporter.py`. This is to make it easy to implement some changes without constantly having to distill the common part out. The switch later will also be easier.